code/core/src/components/components/FocusRing/FocusRing.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useEffect, useRef, useState } from 'react';
+
+import { styled } from 'storybook/theming';
+
+import { useLocationHash } from '../shared/useLocationHash';
+
+export const FocusOutline = styled.div<{ active?: boolean; outlineOffset?: number }>(
+  ({ theme, active = false, outlineOffset = 0 }) => ({
+    width: '100%',
+    borderRadius: 'inherit',
+    transition: 'outline-color var(--transition-duration, 0.2s)',
+    outline: `2px solid ${active ? theme.color.secondary : 'transparent'}`,
+    outlineOffset,
+  })
+);
+
+export const FocusProxy = styled(FocusOutline)<{ targetId: string }>(({ theme, targetId }) => ({
+  [`&:has(#${targetId}:focus-visible)`]: {
+    outlineColor: theme.color.secondary,
+  },
+}));
+
+export const FocusRing = ({
+  active = false,
+  highlightDuration,
+  nodeRef,
+  ...props
+}: React.ComponentProps<typeof FocusOutline> & {
+  highlightDuration?: number;
+  nodeRef?: React.RefObject<HTMLDivElement>;
+}) => {
+  const [visible, setVisible] = useState(active);
+
+  useEffect(() => {
+    setVisible(active);
+    const timeout = setTimeout(setVisible, highlightDuration, false);
+    return () => clearTimeout(timeout);
+  }, [active, highlightDuration]);
+
+  return <FocusOutline {...props} active={highlightDuration ? visible : active} ref={nodeRef} />;
+};
+
+export const FocusTarget = ({
+  targetHash,
+  highlightDuration,
+  ...props
+}: Omit<React.ComponentProps<typeof FocusRing>, 'active'> & {
+  targetHash: string;
+}) => {
+  const nodeRef = useRef<HTMLDivElement>(null);
+  const locationHash = useLocationHash();
+  const [active, setActive] = useState(locationHash === targetHash);
+
+  useEffect(() => {
+    let timeout: NodeJS.Timeout;
+
+    setActive(false);
+    if (locationHash === targetHash) {
+      setTimeout(() => {
```

Comment:
I am not entirely sure whether I understand the usage of `setTimeout` here. Do you now want to use `useLayoutEffect` rather than `useEffect` to access `nodeRef.current`?